R/add_ci.R

```diff
@@ -404,3 +478,202 @@ calculate_prop_ci <- function(x, n, statistic, method, conf.level, style_fun) {
+      tbl$meta_data %>%
+      filter(.data$variable %in% .env$variable) %>%
+      purrr::pluck("df_stats", 1) %>%
+      dplyr::rowwise() %>%
+      mutate(
+        ci = calculate_svyprop_ci(
+          variable = .data$variable,
+          variable_levels = .data$variable_levels,
+          by = .env$by,
+          level = .data[["by"]],
+          tbl = .env$tbl,
+          method = svyprop_method,
+          statistic = statistic[[variable]],
+          conf.level = conf.level,
+          style_fun = style_fun[[variable]],
+          df = df
+        )
+      )
+  }
+  else {
+    glue("Error with variable '{variable}'. Method '{method[[variable]]}' ",
+         "cannot be applied to summary type '{summary_type[[variable]]}'.") %>%
+      stop(call. = FALSE)
+  }
+
+  df_single_ci %>%
+    tidyr::pivot_wider(
+      id_cols = any_of("variable_levels"),
+      values_from = "ci",
+      names_from = "col_name"
+    ) %>%
+    select(all_stat_cols()) %>%
+    dplyr::rename_with(.fn = ~ paste0(., "_ci"))
+}
+
+calculate_svymean_ci <- function(variable, by, level, tbl,
+                                 statistic, conf.level, style_fun, df) {
+  if (is.null(by) || is.na(level))
+    design <- tbl$inputs$data
+  else
+    design <- subset(
+      tbl$inputs$data,
+      tbl$inputs$data$variables[[by]] == level
+    )
+
+  df_ci <-
+    survey::svymean(
+      c_form(right = variable),
+      design = design,
+      na.rm = TRUE
+    ) %>%
+    stats::confint(level = conf.level, df = df) %>%
+    dplyr::as_tibble() %>%
+    set_names(c("conf.low", "conf.high"))
+
+  # round and format CI
+  df_ci %>%
+    select(all_of(c("conf.low", "conf.high"))) %>%
+    dplyr::mutate_all(style_fun) %>%
+    glue::glue_data(statistic) %>%
+    as.character()
+}
+
+calculate_svymedian_ci <- function(variable, by, level, tbl, method,
+                                 statistic, conf.level, style_fun, df) {
+  if (is.null(by) || is.na(level))
+    design <- tbl$inputs$data
+  else
+    design <- subset(
+      tbl$inputs$data,
+      tbl$inputs$data$variables[[by]] == level
+    )
+
+  df_ci <-
+    survey::svyquantile(
+      c_form(right = variable),
+      design = design,
+      quantiles = .5,
+      alpha = 1 - conf.level,
+      interval.type = method,
+      ci = TRUE,
+      na.rm = TRUE,
+      df = df
+    ) %>%
+    purrr::pluck(1) %>%
+    dplyr::as_tibble() %>%
+    set_names(c("estimate", "conf.low", "conf.high", "se"))
+
+  # round and format CI
+  df_ci %>%
+    select(all_of(c("conf.low", "conf.high"))) %>%
+    dplyr::mutate_all(style_fun) %>%
+    glue::glue_data(statistic) %>%
+    as.character()
+}
+
+calculate_svyprop_ci <- function(variable, variable_levels, by, level, tbl,
+                                 method, statistic, conf.level, style_fun, df) {
+  if (is.null(by) || is.na(level))
+    design <- tbl$inputs$data
+  else
+    design <- subset(
+      tbl$inputs$data,
+      tbl$inputs$data$variables[[by]] == level
+    )
+
```

Comment:
potential catch for dichotomous variables, to allow `calculate_svyprop_ci()` to run 
```r
if (!exists("tbl$meta_data$variable_levels") && tbl$meta_data$summary_type == "dichotomous") {
    variable_levels <- unlist(tbl$meta_data$dichotomous_value)
  }
```